cpp/inc/bond/core/select_protocol.h

```diff
@@ -8,203 +8,224 @@
 #include "runtime_schema.h"
 #include "exception.h"
 
+#ifdef BOND_NO_CXX14_GENERIC_LAMBDAS
+#include <functional>
+#endif
+
+
 namespace bond
 {
 namespace detail
 {
 
 
 // Overload of Apply used to extract bonded<T> from marshaled payload
-template <typename T, typename U, typename Reader>
+template <typename Protocols, typename T, typename U, typename Reader>
 inline bool
 Apply(const boost::reference_wrapper<bonded<T> >& ref, const bonded<U, Reader>& value)
 {
-    value.Deserialize(ref.get());
+    value.template Deserialize<Protocols>(ref.get());
     return false;
 }
 
 
-// Select protocol and apply transform using compile-time schema
-template <typename T, typename Buffer, typename Transform, typename EndIter>
-inline std::pair<ProtocolType, bool> NextProtocol(
-    const EndIter&,
-    const EndIter&,
-    Buffer&,
-    const Transform&)
+template <typename Protocols, typename F>
+inline auto TryEachProtocol(F&& f)
+#ifdef BOND_NO_CXX14_RETURN_TYPE_DEDUCTION
+    -> decltype(mpl::try_apply<typename Protocols::type>(std::forward<F>(f)))
+#endif
 {
-    UnknownProtocolException();
-    return std::make_pair(MARSHALED_PROTOCOL, false);
+    return mpl::try_apply<typename Protocols::type>(std::forward<F>(f));
 }
 
 
-template <typename T, typename Buffer, typename Transform, typename Iter, typename EndIter>
-inline std::pair<ProtocolType, bool> NextProtocol(
-    const Iter&,
-    const EndIter& end,
-    Buffer& input,
-    const Transform& transform)
+template <typename T, typename Protocols>
+struct NextProtocolFunctor
 {
-    typedef typename boost::mpl::deref<Iter>::type Reader;
+    template <typename Buffer, typename Transform, typename Reader>
+    inline boost::optional<std::pair<ProtocolType, bool> >
+    operator()(Buffer& input, const Transform& transform, const mpl::identity<Reader>&) const
+    {
+        Reader reader(input);
 
-    Reader reader(input);
+        if (reader.ReadVersion())
+        {
+            return std::make_pair(
+                static_cast<ProtocolType>(Reader::magic),
+                Apply<Protocols>(transform, bonded<T, ProtocolReader>(reader)));
+        }
 
-    if (reader.ReadVersion())
-    {
-        return std::make_pair(
-            static_cast<ProtocolType>(Reader::magic), 
-            Apply(transform, bonded<T, ProtocolReader>(reader)));
+        return {};
```

Comment:
Trying to avoid `boost::none` because it is going to change with `std::optional` (it uses `nullopt_t`) and usually preferring `{}` as it is more universal. But will change to `boost::none` if you still think it is cleaner.